src/subhalo_tracking.cpp

```diff
@@ -175,6 +176,151 @@ inline HBTInt GetLocalHostId(HBTInt pid, const HaloSnapshot_t &halo_snap, const
   }
   return hostid;
 }
+
+/* Identify and store the particle IDs of the most bound MinNumTracerPartOfSub
+ * collisionless tracers. Used to identify host FOF groups. */
+void GetTracerIds(vector<HBTInt>::iterator particle_ids, const Subhalo_t &Subhalo)
+{
+  /* Initialise vector. This will make it so the code knows when to stop looking
+   * for tracers, since orphans will have all but the first with NullParticleId */
+  fill(particle_ids, particle_ids + HBTConfig.MinNumTracerPartOfSub, SpecialConst::NullParticleId);
+
+  /* Iterate over the particle list to find tracers. */
+  int BoundRanking = 0;
+  for (auto particle = Subhalo.Particles.begin(); particle != Subhalo.Particles.begin() + Subhalo.Nbound; particle++)
+  {
+    if (particle->IsTracer()) // Only use tracers.
+      particle_ids[BoundRanking++] = particle->Id;
+
+    if (BoundRanking == HBTConfig.MinNumTracerPartOfSub) // Use up to a user-defined number of tracers.
+      break;
+  }
+}
+
+/* Identify which FOF is host to the particles. If we have a value of -2, that
+ * means the particle was not found in the particle information available to the
+ * local rank. */
+bool GetTracerHosts(vector<HBTInt>::iterator particle_hosts, vector<HBTInt>::const_iterator particle_ids,
+                    const HaloSnapshot_t &halo_snap, const ParticleSnapshot_t &part_snap)
+{
+  /* Initialise vector. This will make it so the code knows when to stop looking
+   * for tracers, since orphans will have all but the first with NullParticleId */
+  fill(particle_hosts, particle_hosts + HBTConfig.MinNumTracerPartOfSub, -2);
+
+  bool MakeDecision = true;
+
+  /* Iterate over the particle list to find hosts. */
+  for (int BoundRanking = 0; BoundRanking < HBTConfig.MinNumTracerPartOfSub; BoundRanking++)
+  {
+    // No more tracers left (should only occur for orphans!)
+    if (particle_ids[BoundRanking] == SpecialConst::NullParticleId)
+      break;
+
+    // Get host, which can be -1
+    particle_hosts[BoundRanking] = halo_snap.ParticleHash.GetIndex(particle_ids[BoundRanking]);
+
+    /* We cannot find the particle in the current rank. We will therefore need to
+     * try to find it in other ranks (and hence delay making a host decision) */
+    if ((particle_hosts[BoundRanking] == -1) &&
+        (part_snap.GetIndex(particle_ids[BoundRanking]) == SpecialConst::NullParticleId))
+    {
+      particle_hosts[BoundRanking]--; // Turns it to -2
+      MakeDecision = false;
```

Comment:
Should I set MakeDecision = false even if we find values of -1? I have found examples where one rank finds most particles to be part of a FOF, whereas another one finds them to be -1. 

I think it has to do with how the the snapshot/halo particles are distributed. Consider a case where the most bound particles of a subgroup are in rank 0, but the FOF (e.g. FOFID = 1) they are hosted in is located in rank 1. Querying the halo distribution in rank 1 would give particle_hosts = 1. Doing so in rank 0 would yield -1, since the halo information is not there, but the particles are. 

Would we therefore still need to check cases where all particle_hosts values = -1?

Reply:
I'm not sure I understand this. If all particle_hosts are -1 doesn't that mean that all of the tracer particle IDs have been found in the halo snapshot and determined to be in no FoF group? If so, I think there's no need to check on other ranks because we definitely wont find those particle IDs anywhere else - there should be no duplicate IDs in the halo snapshot.

Surely each element in particle_hosts has to be -2 on all ranks except one? I don't see how you can have particle_hosts[i]=-1 on one rank and >=0 on another.

Reply:
The particles of one subhalo might be split between different FoFs though, or some in a FoF and some in no FoF. Then the host identified by the highest score would differ between MPI ranks. But that can still only happen if you have at least one particle_hosts=-2, so I think the code is ok as it is?

Reply:
Oh, but I'm missing the complication that the halo snapshot and particle snapshot aren't distributed the same way.

Reply:
So we can't tell the difference between "this particle is in no FoF so no need to look elsewhere" and "this particle is not in the halosnap on this rank so we have to search other ranks"?

Reply:
As far as I understand, I do not think we can disentangle those two scenarios. The code essentially checks if the particle is part of a FOF hosted in the present rank ( `>= 0` if so, `-1` if not), and whether it can be found in the particle distribution in the task (`-2` if not, `-1` if so). Thinking of a worst case scenario, we could have the host FOF (e.g. `7`) of the 10 tracers in rank 0, but the actual particles in rank 1.  

The code would then give for rank 0:

`7 7 7 7 7 7 7 7 7 7`

for rank 1:

`-1 -1 -1 -1 -1 -1 -1 -1 -1 -1`

for rank 2:

`-2 -2 -2 -2 -2 -2 -2 -2 -2 -2`

The outcome would then be dependent on which rank the subhalo is located in. We can only be sure we have the correct information if all the numbers are not equal to `-1` or `-2`

Reply:
In any case, even if we find a host of `-1`, the code will still try to obtain information from other ranks. See https://github.com/SWIFTSIM/HBTplus/blob/30f920f76e72900da9ccf470f75074c617e0fe11/src/subhalo_tracking.cpp#L357. That line is selecting which subhaloes need information from extra tasks.

Reply:
For your second comment, I had a (uncommitted) print statement that would complain if we found different target ranks for particles within the same subgroup. I do not think I ever saw it trigger, but perhaps I should check it again to confirm this is the case.

Reply:
Lastly, regardless of my comment concerning https://github.com/SWIFTSIM/HBTplus/blob/30f920f76e72900da9ccf470f75074c617e0fe11/src/subhalo_tracking.cpp#L357
I think we should still keep the boolean flag `MakeDecision = false`, if we have a single particle with a value of `-1`. This is because the particle might be associated to a subgroup in a different rank, and hence tip the balance towards one candidate or another. If we remove the flag, we might be making a premature decision.

Reply:
Ah, good catch. I put that in place before realising that -1 would still require further looking. Let me change that now.